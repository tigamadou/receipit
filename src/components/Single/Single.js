import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import Axios from 'axios';
import { toast } from 'react-toastify';
import { useParams } from 'react-router-dom';
import { connect } from 'react-redux';
import { createRecipe, setHeader } from '../../redux/actions/index';

const getData = (id, recipes) => recipes.filter((data) => data.idMeal === id);

const Single = ({
  recipes, element, createRecipe, setHeader,
}) => {
  const [data, setData] = useState(element);
  const { id } = useParams();
  const ingredients = [];
  for (let i = 0; i < 20; i += 1) {
    ingredients.push(i);
  }
  const getRecipe = () => {
    Axios.get(`https://www.themealdb.com/api/json/v1/1/lookup.php?i=${id}`)
      .then((response) => {
        setData(response.data.meals[0]);
        createRecipe(response.data.meals[0]);
      })
      .catch(() => {
        toast.error('Oups! Something went wrong. Please try again later.', {
          position: 'top-right',
          autoClose: 5000,
          hideProgressBar: false,
          closeOnClick: true,
          pauseOnHover: true,
          draggable: true,
          progress: undefined,
        });
      });
  };

  useEffect(() => {
    const data = getData(id, recipes)[0];
    if (data) {
      setData(data);
      return;
    }
    getRecipe();
  }, []);

  useEffect(() => {
    if (data) {
      setHeader({ title: data.strMeal, back: `/category/${data.strCategory}`, search: true });
    }
  }, [data]);
  return (
    <div className="container">
      {data
        && (
        <div>
          <img src={data.strMealThumb} alt={data.strMeal} />
          <h1>{data.strMeal}</h1>
          <div>{data.strArea}</div>
          <div>{data.strCategory}</div>
          <div>{data.strCategory}</div>
          <div>
            <h3>Ingredients</h3>
            <ul>
              { ingredients.map((e) => {
                if (data[`strIngredient${e + 1}`] !== '' && data[`strIngredient${e + 1}`] !== null) {
                  return <li key={e}>{ data[`strIngredient${e + 1}`] }</li>;
                }
                return <></>;
              })}
            </ul>
          </div>
          <div>
            <h3>Instruction</h3>
            <p>
              {data.strInstructions}
            </p>
          </div>
        </div>
        )}
    </div>
  );
};

Single.defaultProps = {
  element: null,
  recipes: null,
  createRecipe: null,
  setHeader: null,
};

Single.propTypes = {
  element: PropTypes.instanceOf(Object),
  recipes: PropTypes.arrayOf(PropTypes.instanceOf(Object)),
  createRecipe: PropTypes.func,
  setHeader: PropTypes.func,
};

const mapDispatchToProps = (dispatch) => ({
  createRecipe: (recipe) => dispatch(createRecipe(recipe)),
  setHeader: (header) => dispatch(setHeader(header)),
});

const mapStateToProps = (state) => ({
  recipes: state.recipes,
});

export default connect(mapStateToProps, mapDispatchToProps)(Single);
