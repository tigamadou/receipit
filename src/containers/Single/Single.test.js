import React from 'react';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import {
  BrowserRouter as Router,
  Route,
} from 'react-router-dom';
import { Provider } from 'react-redux';
import Single from './Single';
import storeReducer from '../../redux/store';

const store = storeReducer();

const data = {
  dateModified: null,
  idMeal: '52874',
  strArea: 'British',
  strCategory: 'Beef',
  strCreativeCommonsConfirmed: null,
  strDrinkAlternate: null,
  strImageSource: null,
  strIngredient1: 'Beef',
  strIngredient2: 'Plain Flour',
  strIngredient3: 'Rapeseed Oil',
  strIngredient4: 'Red Wine',
  strIngredient5: 'Beef Stock',
  strIngredient6: 'Onion',
  strIngredient7: 'Carrots',
  strIngredient8: 'Thyme',
  strIngredient9: 'Mustard',
  strIngredient10: 'Egg Yolks',
  strIngredient11: 'Puff Pastry',
  strIngredient12: 'Green Beans',
  strIngredient13: 'Butter',
  strIngredient14: 'Salt',
  strIngredient15: 'Pepper',
  strIngredient16: '',
  strIngredient17: '',
  strIngredient18: '',
  strIngredient19: '',
  strIngredient20: '',
  strInstructions: 'Preheat the oven to 150C/300F/Gas 2',
  strMeal: 'Beef and Mustard Pie',
  strMealThumb: 'https://www.themealdb.com/images/media/meals/sytuqu1511553755.jpg',
  strMeasure1: '1kg',
  strMeasure2: '2 tbs',
  strMeasure3: '2 tbs',
  strMeasure4: '200ml',
  strMeasure5: '400ml',
  strMeasure6: '1 finely sliced',
  strMeasure7: '2 chopped',
  strMeasure8: '3 sprigs',
  strMeasure9: '2 tbs',
  strMeasure10: '2 free-range',
  strMeasure11: '400g',
  strMeasure12: '300g',
  strMeasure13: '25g',
  strMeasure14: 'pinch',
  strMeasure15: 'pinch',
  strMeasure16: '',
  strMeasure17: '',
  strMeasure18: '',
  strMeasure19: '',
  strMeasure20: '',
  strSource: 'https://www.bbc.co.uk/food/recipes/beef_and_mustard_pie_58002',
  strTags: 'Meat,Pie',
  strYoutube: 'https://www.youtube.com/watch?v=nMyBC9staMU',

};
describe('Single component tests', () => {
  test('renders single component ', () => {
    // eslint-disable-next-line max-len
    render(<Provider store={store}><Router><Route><Single element={data} /></Route></Router></Provider>);
    // eslint-disable-next-line no-unused-expressions
    expect(screen.getByText(data.strMeal)).toBeInTheDocument;
  });

  test('should match with snapshot', () => {
    const tree = renderer
      // eslint-disable-next-line max-len
      .create(<Provider store={store}><Router><Route><Single element={data} /></Route></Router></Provider>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
