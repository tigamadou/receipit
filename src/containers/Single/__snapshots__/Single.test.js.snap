// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single component tests should match with snapshot 1`] = `
<div
  className="container"
>
  <div
    className="single"
  >
    <div
      className="tabs"
    >
      <a
        aria-current="page"
        className="white is_sm active"
        href="//"
        onClick={[Function]}
        style={Object {}}
      >
        Preview
      </a>
      <a
        aria-current={null}
        className="white is_sm"
        href="//ingredients/"
        onClick={[Function]}
      >
        Ingredients
      </a>
      <a
        aria-current={null}
        className="white is_sm"
        href="//preparation/"
        onClick={[Function]}
      >
        Preparation
      </a>
    </div>
    <div
      className="details"
      style={
        Object {
          "backgroundImage": "url(https://www.themealdb.com/images/media/meals/sytuqu1511553755.jpg)",
        }
      }
    >
      <div
        className="preview"
      >
        <div
          className="overlay_"
        >
          <h1
            className="white is_default"
          >
            Beef and Mustard Pie
          </h1>
          <div
            className="primary"
          >
            British
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
